File: Tests/AINotebookCoreTests/SystemPromptTests.swift
import XCTest
@testable import AINotebookCore

final class SystemPromptTests: XCTestCase {

    func testRendersHitsAsNumberedBlocks() {
        let hits = [
            RetrievalHit(chunkId: 10, sourceId: 1, score: 0.9, snippet: "alpha facts"),
            RetrievalHit(chunkId: 11, sourceId: 2, score: 0.7, snippet: "beta facts")
        ]
        let prompt = SystemPrompt.compose(hits: hits)
        XCTAssertTrue(prompt.contains("[1] alpha facts"))
        XCTAssertTrue(prompt.contains("[2] beta facts"))
    }

    func testIncludesCitationInstruction() {
        let prompt = SystemPrompt.compose(hits: [])
        XCTAssertTrue(prompt.lowercased().contains("cite"))
        XCTAssertTrue(prompt.contains("[N]"))
    }

    func testNoHitsStillProducesValidPrompt() {
        let prompt = SystemPrompt.compose(hits: [])
        XCTAssertFalse(prompt.isEmpty)
    }
}

File: windows/tests/AINotebook.Core.Tests/Rag/SystemPromptTests.cs
using AINotebook.Core.Models;
using AINotebook.Core.Rag;
using Xunit;

namespace AINotebook.Core.Tests.Rag;

public class SystemPromptTests
{
    // SystemPromptTests.testRendersHitsAsNumberedBlocks
    [Fact]
    public void RendersHitsAsNumberedBlocks()
    {
        var hits = new[]
        {
            new RetrievalHit(1, 1, 0.5f, "alpha facts"),
            new RetrievalHit(2, 1, 0.4f, "beta facts"),
        };
        var prompt = SystemPrompt.Compose(hits);
        Assert.Contains("[1] alpha facts", prompt);
        Assert.Contains("[2] beta facts", prompt);
    }

    // SystemPromptTests.testIncludesCitationInstruction
    [Fact]
    public void IncludesCitationInstruction()
    {
        var prompt = SystemPrompt.Compose(Array.Empty<RetrievalHit>());
        Assert.Contains("cite", prompt.ToLowerInvariant());
        Assert.Contains("[N]", prompt);
    }

    // SystemPromptTests.testNoHitsStillProducesValidPrompt
    [Fact]
    public void NoHitsStillProducesNonEmptyPrompt() =>
        Assert.False(string.IsNullOrEmpty(SystemPrompt.Compose(Array.Empty<RetrievalHit>())));

    // Includes note section when provided (ChatEngineCurrentNoteContextTests companion).
    [Fact]
    public void IncludesNoteSectionWhenProvided()
    {
        var prompt = SystemPrompt.Compose(Array.Empty<RetrievalHit>(), "flour 500g");
        Assert.Contains("CURRENTLY OPEN NOTE", prompt);
        Assert.Contains("flour 500g", prompt);
    }

    [Fact]
    public void OmitsNoteSectionWhenNullOrBlank()
    {
        Assert.DoesNotContain("CURRENTLY OPEN NOTE", SystemPrompt.Compose(Array.Empty<RetrievalHit>(), null));
        Assert.DoesNotContain("CURRENTLY OPEN NOTE", SystemPrompt.Compose(Array.Empty<RetrievalHit>(), "   "));
    }
}
